src/UI/FlatList.re

```diff
@@ -4,106 +4,179 @@
  * Virtualized list helper
  */
 
+open Revery;
 open Revery.UI;
+open Revery_UI_Components;
 
 type renderFunction = int => React.syntheticElement;
 
-let component = React.component("FlatList");
-
-let additionalRowsToRender = 1;
-
-type glRenderFunction = (int, float) => unit;
-
-let render =
-    (~scrollY=0., ~rowHeight, ~height, ~count, ~render: glRenderFunction, ()) => {
-  let rowsToRender = rowHeight > 0. ? int_of_float(height /. rowHeight) : 0;
-  let startRowOffset =
-    rowHeight > 0. ? int_of_float(scrollY /. rowHeight) : 0;
-  let pixelOffset = mod_float(scrollY, rowHeight);
+module Constants = {
+  let additionalRowsToRender = 1;
+  let scrollBarThickness = 6;
+  let scrollTrackColor = Color.rgba(0.0, 0.0, 0.0, 0.4);
+  let scrollThumbColor = Color.rgba(0.5, 0.5, 0.5, 0.4);
+};
 
-  let i = ref(max(startRowOffset - additionalRowsToRender, 0));
+module Styles = {
+  let container = (~width as w, ~height as h) =>
+    Style.[
+      position(`Relative),
+      top(0),
+      left(0),
+      width(w),
+      height(h),
+      overflow(`Hidden),
+    ];
+
+  let slider =
+    Style.[
+      position(`Absolute),
+      right(0),
+      top(0),
+      bottom(0),
+      width(Constants.scrollBarThickness),
+    ];
+
+  let viewport = (~isScrollbarVisible) =>
+    Style.[
+      position(`Absolute),
+      top(0),
+      left(0),
+      right(isScrollbarVisible ? 0 : Constants.scrollBarThickness),
+    ];
+
+  let item = (~offset, ~rowHeight) =>
+    Style.[
+      position(`Absolute),
+      top(offset),
+      left(0),
+      right(0),
+      height(rowHeight),
+    ];
+};
 
-  let len = count;
+let component = React.component("FlatList");
 
-  while (i^ < rowsToRender
-         + additionalRowsToRender
-         + startRowOffset
-         && i^ < len) {
-    let item = i^;
-    let rowOffset = float_of_int(item - startRowOffset) *. rowHeight;
+let render = (~menuHeight, ~rowHeight, ~count, ~scrollTop, ~renderItem) =>
+  if (rowHeight <= 0) {
+    [];
+  } else {
+    let startRow = scrollTop / rowHeight;
+    let startY = scrollTop mod rowHeight;
+    let rowsToRender =
+      min(
+        menuHeight / rowHeight + Constants.additionalRowsToRender,
+        count - startRow,
+      );
+    let indicesToRender = List.init(rowsToRender, i => i + startRow);
```

Comment:
I think the crash I saw is related to this. If I log out the `rowsToRender` value, I see that sometimes it can become negative:
```
[INFO] rowsToRender: -3
Fatal error: exception Invalid_argument("List.init")
```

Might be sufficient to set it up so that if `rowsToRender` is less than or equal to zero, we just return an empty list?

Reply:
Ah, right, there's some flawed assumptions here, such as the selected index always being valid, which doesn't hold because the items array might change without the selection being adjusted. I also think `selected` might be negative to indicate no selection, and that this causes problems somewhere.

It's an easy fix, but there might be a few more of these to weed out, that I've fixed in a later commit in the original PR.

Reply:
Should be fixed now. Also backported the other fixes I could find in FlatList